Require tlog_id when log_id_ranges is passed in

lkatalin opened the pull request:
#### Summary
Based on the conversation @priyawadhwa in #727 , this separates concerns more cleanly between the `log_id_ranges` flag and `tlog_id` flag. `tlog_id` now specifies the active tree (shard) and can be used by itself. `log_id_ranges` specifies inactive trees (shards) and must be used in conjunction with `tlog_id`.

priyawadhwa commented on cmd/rekor-server/app/flags.go:
```diff
@@ -27,41 +27,43 @@ type LogRangesFlag struct {
 
+	// The last entry is special and should not have a terminating range, because this is active.
+	// It is specified by the tlog_id.
 	inputRanges = append(inputRanges, sharding.LogRange{
```
Instead of appending the active ID, we can add a `ActiveTreeID int` field to the `LogRangesFlag` struct that'll make it easier to access it

lkatalin replied:
@priyawadhwa That's a good idea. This also implies we may have to move methods like `func (l *LogRanges) ResolveVirtualIndex(index int)` and `func (l *LogRanges) TotalLength() int64` to be methods on `logRangesFlag` instead of `LogRanges` in order to work across all shards (active + inactive), does that seem reasonable?

lkatalin replied:
I've actually got a better way of doing it, I'm turning the `LogRanges` struct into this:
```
type LogRanges struct {
        inactive []LogRange
        active   int64
}
```
That should be the simplest way, stay tuned!

priyawadhwa replied:
awesome!

priyawadhwa commented on cmd/rekor-server/app/serve.go:
```diff
@@ -103,11 +103,15 @@ var serveCmd = &cobra.Command{
 		if rangeMap != "" {
-			if err := logRangeMap.Set(rangeMap); err != nil {
-				log.Logger.Fatal("unable to set logRangeMap from flag: %v", err)
+			if tlogID == 0 {
```
we don't need this check because we check for it later in `logRangeMap.Set` again

lkatalin committed:
Require tlog_id when inactive shard config file is passed in

tlog_id specifes the active shard and is kept for backwards compatibility.
To avoid replicating information, the shard config file is used only to
specify inactive shards and must be used in conjunction with a tlog_id flag.
Together, these build the logRanges type in the sharding module.

Signed-off-by: Lily Sturmann <lsturman@redhat.com>

priyawadhwa commented on cmd/rekor-server/app/root.go:
```diff
@@ -63,7 +63,7 @@ func init() {
 	rootCmd.PersistentFlags().String("trillian_log_server.address", "127.0.0.1", "Trillian log server address")
 	rootCmd.PersistentFlags().Uint16("trillian_log_server.port", 8090, "Trillian log server port")
-	rootCmd.PersistentFlags().Uint("trillian_log_server.tlog_id", 0, "Trillian tree id")
+	rootCmd.PersistentFlags().String("trillian_log_server.tlog_id", "0", "Trillian tree id")
```
sorry I wasn't clear when we chatted offline, I think this should actually remain a `Uint` for backwards compatibility.

lkatalin replied:
Aaahhh, got it. No worries, easy to fix.

priyawadhwa commented on pkg/sharding/log_index.go:
```diff
@@ -16,18 +16,29 @@ package sharding
 
 	var virtualIndex int64
-	for _, r := range ranges.GetRanges() {
+	// TODO: Are the tree IDs guaranteed to be ordered deterministically in the Ranges?
```
yup, we can assume the user correctly set up the config so that the shards exist in order

lkatalin replied:
Perfect. Thank you!!

priyawadhwa commented on pkg/sharding/ranges.go:
```diff
@@ -44,6 +45,9 @@ func NewLogRanges(path string, treeID string) (LogRanges, error) {
 		return LogRanges{}, errors.Wrapf(err, "%s is not a valid int64", treeID)
 	}
+	if id == 0 {
+		return LogRanges{}, errors.New("non-zero tlog_id required when passing in shard config filepath")
```
We could add the fix to this error message to make it easier for users:

```
please set the active tree ID via the `--trillian_log_server.tlog_id` flag
```

priyawadhwa commented on pkg/sharding/ranges.go:
```diff
@@ -53,59 +57,76 @@ func NewLogRanges(path string, treeID string) (LogRanges, error) {
+	return l.inactive == nil
 }
 
 func (l *LogRanges) Empty() bool {
```
i think this might be unused at this point and can be removed?

lkatalin replied:
Sounds good.

priyawadhwa reviewed (commented):
Looks great!

priyawadhwa reviewed (approved):
LGTM!
